code/modules/reagents/chemistry/holder.dm

```diff
@@ -270,24 +270,24 @@ var/const/INJECT = 5 //injection
 			for(var/addiction in addiction_list)
 				var/datum/reagent/R = addiction
 				if(M && R)
-					if(R.addiction_stage <= 0)
-						R.addiction_stage++
-					if(R.addiction_stage > 0 && R.addiction_stage <= 10)
-						R.addiction_act_stage1(M)
-						R.addiction_stage++
-					if(R.addiction_stage > 10 && R.addiction_stage <= 20)
-						R.addiction_act_stage2(M)
-						R.addiction_stage++
-					if(R.addiction_stage > 20 && R.addiction_stage <= 30)
-						R.addiction_act_stage3(M)
-						R.addiction_stage++
-					if(R.addiction_stage > 30 && R.addiction_stage <= 40)
-						R.addiction_act_stage4(M)
-						R.addiction_stage++
-					if(R.addiction_stage > 40)
-						M << "<span class='notice'>You feel like you've gotten over your need for [R.name].</span>"
-						addiction_list.Remove(R)
+					R.addiction_stage++
+					switch(R.addiction_stage)
+						if(1 to 10)
+							need_mob_update += R.addiction_act_stage1(M)
+						if(10 to 20)
+							need_mob_update += R.addiction_act_stage2(M)
+						if(20 to 30)
+							need_mob_update += R.addiction_act_stage3(M)
+						if(30 to 40)
+							need_mob_update += R.addiction_act_stage4(M)
+						if(40 to INFINITY)
+							M << "<span class='notice'>You feel like you've gotten over your need for [R.name].</span>"
+							addiction_list.Remove(R)
 		addiction_tick++
+	if(M && need_mob_update)
+		M.updatehealth()
+		M.update_canmove()
+		M.update_stamina()
```

Comment:
If one of the metabolized reagent changed our mob's health or stat or mobility or stamina, we update all these here and only once, instead of multiple times inside each reagent's on_mob_life().

I didn't think necessary to separate the possible updates (i.e. if(updatehealth_needed) , if(update_stamina_needed), etc). Having on_mob_life() return a list of all the type of updates needed seemed a bit too much.